### PR TITLE
Sql repo experiment

### PR DESCRIPTION
**Pull request by jeromekelleher**
This is a POC/WIP towards using a SQLite DB as the data repository. After looking at the upcoming requirements for URL based access to data files, metadata improvements to repo manager and scalability requirements for serving 1000 Genomes data, it seemed quicker to just tackle the full DB based repo rather than put more and more bandages on the directory-tree based repo concept.

There are some fairly radical changes here, but I don't think they are too intrusive as a whole. I've tried to minimise the parts of the code that we touch here, and particularly avoid all protobuf-affected code as much as possible. There are several more fairly hefty refactors that we could do after this which would remove a lot of code and complexity (i.e., by using SQL to make paging easier).

## Basic Idea

We get rid of the idea of a hierarchy of files representing the data repository and replace it with a single SQLlite DB. Within this DB we have tables for Datasets, ReadGroupSets, ReadGroups and all the other container objects. These are related to each other via IDs in usual relational manner. Everything that depends on a dataUrl (like a ReadGroupSet) stores the path/URL of the file in question in the DB. The storage/locations of the files is entirely up to the repo administrator. This simplifies many things, particularly the logic in repo manager, as there is no need for copy/move/link semantics any more (effectively, we always link now).

## Transitional Design

The current approach is a transitional design, as we are not fully taking advantage of SQL when running our queries. This is a deliberate approach for now to get things running without breaking too much along the way. However, ultimately we would want to remove the bit where we ``load()`` everything from the DB into memory, and just work directly from the database's model of the world. This would have many advantages such as the possibility of online updating of data, and super-fast server startup times.

## Simplification of repo manager CLI

I have made a start on making changes to the repo manager CLI, which basically involves taking all of the functionality in the current ``repo_manager`` module and spreading this between the immediate CLI interface and the ``datarepo`` module. I think this simplifies adding more functionality to the CLI that we need and keeps the functionality of updating and checking the state of the repository in one place. See the updates to the cli for ``add-referenceset`` as an example. This is rough so far, but the basic idea is there. I know the cli module is getting too big, but we'll be refactoring this into the cli package soon (see #934), and it's good to keep all the CLI related code together.

## Impact for users

Users will have to fully import their data. If there are any, this will be a disruptive change. However, from a user-demo perspective, little will change. We can update the ga4gh-example-data to have the repo DB file in the root directory and contain paths to all the relevant files.

## Repo API

Here's some client code for the repo API:
```python
    repo = datarepo.SqlDataRepository("example-data.db")
    if repo.exists():
        repo.delete()
    
    # Creation and update semantics; TODO create a context manager?
    repo.initialise()
    repo.openDb()
    
    fastaFile = "GRCH38.fa.gz"
    rs1 = references.HtslibReferenceSet("GRCH38", fastaFile)
    repo.insertReferenceSet(rs1)

    ds1 = datasets.Dataset("ds1", "Test dataset 1")
    repo.insertDataset(ds1)

    bamFile = "ga4gh-example-data/datasets/1kg-p3-subset/reads/HG00096.bam"
    readGroupSet = reads.HtslibReadGroupSet(
            ds1, "HG00096", bamFile, bamFile + ".bai")
    repo.insertReadGroupSet(readGroupSet)
    repo.closeDb()

    # Normal read-only usage.
    repo.load()
```

Consider ``HtslibReadGroupSet``. This is created using a name, BAM URL and an index file. It then has two methods, ``populateFromFile`` and ``populateFromRow``. The idea then is that when we are inserting the ReadGroupSet into the DB we call ``populateFromFile`` to read all the headers etc. When we are just creating the object at server run time, we call ``populateFromRow`` which updates the instance variables of the object based on the values from the DB. This means that we _do not_ need to read the data file at run time unless we are actually retrieving read data. This will make a big difference when we're dealing with thousands of remote files in 1000G.

## TODO

1. Finish reference handling, and perform basic checks when importing referenceSets (currently assuming a single FASTA as it's quite a bit simpler, #955)
2. Add in tables for ReadStats and Programs.
3. Add support for Variants (VariantSets, CallSets, VariantAnnotationSets).
4. Fix tests.
5. Update the download_data script to use the new 'repo' API to make creating the example data easier.

In summary, this is an alternative proposal to #1051, and also tackles a number of other issues at the same time. I think I could probably get the basic framework done in another day. As I say, this is a fairly major change so please review what I've done so far and weigh in with what you think (if you think it's rubbish, say so!). Again, apologies for the size of the change but I couldn't see any other way to meet all of our upcoming requirements without tacking this.

### CODE REVIEW

**Comment by david4096** [comment]
@jeromekelleher this is stellar!

**Comment by macieksmuga** [comment]
@jeromekelleher +10000 for getting this ball rolling! Will go through the details shortly.

File: ga4gh/datamodel/reads.py [hunk 1]
```diff
@@ -553,17 +573,29 @@ def __init__(self, parentContainer, localId, readGroupHeader=None):
...
+        self._experimentDescription = readGroupHeader.get('DS', None)
+        self._library = readGroupHeader.get('LB', None)
+        self._platformUnit = readGroupHeader.get('PU', None)
+        self._runTime = readGroupHeader.get('DT', None)
+
+    def populateFromRow(self, row):
+        """
+        Populate the instance variables using the specified DB row.
+        """
+        self._sampleId = row[b'sampleId']
```
Comment (david4096) [inline ga4gh/datamodel/reads.py]: Excellent, this is exactly what is needed to be able to handle the variety of linkages that support more complicated queries.

File: ga4gh/datamodel/reads.py [hunk 2]
```diff
@@ -242,7 +256,7 @@ def _referenceSetInit(self, dataRepository, shouldThrowExceptions):
             elif referenceSetName != name:
                 if shouldThrowExceptions:
                     raise exceptions.MultipleReferenceSetsInReadGroupSet(
-                        self._samFilePath, name, referenceSetName)
+                        self._dataUrl, name, referenceSetName)
```
Comment (david4096) [inline ga4gh/datamodel/reads.py]: It makes sense to me to do away with this function (`referenceSetInit`) now that we have a DB row for the readgroupset. Reading the reference set name from the file internals is an implicit assumption we probably shouldn't make. When initializing a readgroupset, providing the referenceSet that it is against would address https://github.com/ga4gh/server/issues/1050 , as well as demonstrate how the other containers would be initialized with a referenceSetId.

**Comment by david4096** [comment]
screen shot 2016-04-07 at 4 02 34 pm

It was helpful to me to see the tables and schemas so we have a clear path to the secondary goal: using queries for paging and querying datamodel objects. I see that you've included `name` in the tables, which is great because our searching-by-name support will improve, same with `sampleId`. The work I've done in the biodata protocol to use the current paging scheme to filter callsets and readgroupsets would be significantly improved if the containers were stored in a table like this.

I believe we should store the referenceSetId alongside the containers for readGroupSets.

With this iteration we can just create a variants table, as annotations and callsets are generated logically from a variantset. 

```
VariantSet
id 			TEXT,
datasetId		TEXT
name 		TEXT,
referenceSetId	TEXT
description 	TEXT,
dataDir 		TEXT,
created		TEXT,
updated		TEXT
callSetIds		TEXT
```

Looking forward to updates!

File: ga4gh/datamodel/reads.py (hunk 2 above)
Comment (jeromekelleher) [inline ga4gh/datamodel/reads.py]: We definitely want to make this more robust. It would be nice if we could detect the reference set by looking at the md5s at import time automatically (it would save the user some effort), but this should probably be pushed up the stack closer to the repo manager UI.

**Comment by jeromekelleher** [comment]
Thanks @macieksmuga and @david4096, it's great to hear you like the idea! OK, I'd better finish it up then...

**Comment by jeromekelleher** [comment]
Update:

- Variants, References, Reads queries all working in the server.

TODO:

- VariantAnnotations, OntologyMaps
- Update repo manager to use the new API
- Update download data script to use new API
- Fix the rest of the tests.
- Add back in support for ReadStats and Experiment.

File: ga4gh/datarepo.py [hunk 3]
```diff
@@ -179,9 +200,457 @@ def __init__(
...
+            metadataJson, urlMapJson))
+        for callSet in variantSet.getCallSets():
+            self.insertCallSet(callSet)
+
+    def _readVariantSetTable(self, cursor):
+        cursor.row_factory = sqlite3.Row
+        cursor.execute("SELECT * FROM VariantSet;")
+        for row in cursor:
+            dataset = self.getDataset(row[b'datasetId'])
+            referenceSet = self.getReferenceSet(row[b'referenceSetId'])
```
Comment (david4096) [inline ga4gh/datarepo.py]: Perfect, solves this long standing issue https://github.com/ga4gh/server/issues/689.

**Comment by jeromekelleher** [comment]
Update:

- ReferenceSets based on single FASTA.
- download_data script uses the repo API, and has creates a flat file structure
- Partial refactor of VariantAnnotations (was hitting a lot of problems I didn't understand, so seemed better to move on). Disabled test cases for VA while it's broken.
- Partial refactor of repo manager CLI code. Disabled on repo manager while work is continuing.

**Comment by jeromekelleher** [comment]
Any feedback here @dcolligan? I'm hoping to merge this into a topic branch in the next day or two (will be closing this PR as it won't be ready to merge to master), so it would be good to get your thoughts.

**Comment by diekhans** [comment]
Overall, I like this a lot.  Random thoughts on SQL backend

 1) registry problem (aka data outside of transactions)Since
    some data is still in files, it would be to have the repo
    manager have a verify function to make sure the

 2) distributing datasets - it will be really handy to be able
    to put out standard data sets, like reference genomes,
    GENCODE, RefSeq as single database file (or database +
    fasta).

 3) hooking into ucsc browser database - this will still require
    some thinking

**Comment by jeromekelleher** [comment]
@diekhans: great! We should try and push this in as soon as we can I think, as I think this will really speed up our progress towards server 1000G and generally having a useful server. I've created a topic branch and opened a PR against it in #1079. Therefore, closing this PR and moving discussion over there.

File: ga4gh/cli.py [hunk 4]
```diff
@@ -1493,17 +1495,17 @@ def configtest_main(parser=None):
 ##############################################################################
 
 
-def getRawInput(displayString):
```
Comment (dcolligan) [inline ga4gh/cli.py]: keep this function... it's mocked in some tests

File: ga4gh/cli.py [hunk 5]
```diff
@@ -1803,6 +1854,7 @@ def getRepoParser():
         subparsers, "init", "Initialize a data repository")
     initParser.set_defaults(runner=InitRunner)
     addRepoArgument(initParser)
+    addForceArgument(initParser)
```
Comment (dcolligan) [inline ga4gh/cli.py]: good idea

File: ga4gh/datarepo.py [hunk 6]
```diff
@@ -179,9 +256,536 @@ def __init__(
...
+    def open(self, mode="r"):
+        """
+        Opens this repo in the specified mode.
+
+        TODO: figure out the correct semantics of this and document
+        the intended future behaviour as well as the current
+        transitional behaviour.
+        """
+        if mode not in [MODE_READ, MODE_WRITE]:
+            raise ValueError("Open mode must be 'r' or 'w'")
```
Comment (dcolligan) [inline ga4gh/datarepo.py]: interpolate the r and w strings here into the msg from MODE_READ and MODE_WRITE

File: tests/datadriven/test_variant_annotations.py [hunk 7]
```diff
@@ -13,6 +13,8 @@
 import ga4gh.datamodel as datamodel
 import ga4gh.datamodel.datasets as datasets
 import ga4gh.datamodel.variants as variants
+import ga4gh.datamodel.references as references
+# import ga4gh.datamodel.ontologies as ontologies
```
Comment (dcolligan) [inline tests/datadriven/test_variant_annotations.py]: take out

File: ga4gh/datamodel/datasets.py [hunk 8]
```diff
@@ -6,25 +6,26 @@
...
 import ga4gh.datamodel.reads as reads
 import ga4gh.datamodel.sequenceAnnotations as sequenceAnnotations
 import ga4gh.datamodel.variants as variants
+import ga4gh.datamodel.ontologies as ontologies
 import ga4gh.exceptions as exceptions
 import ga4gh.protocol as protocol
 
 
-class AbstractDataset(datamodel.DatamodelObject):
+class Dataset(datamodel.DatamodelObject):
```
Comment (dcolligan) [inline ga4gh/datamodel/datasets.py]: why the name change?

**Comment by dcolligan** [comment]
Definitely something we need going forward.  Commenting all those tests out does worry me, but I since this is a WIP I take it those will be resolved before this is committed...

File: ga4gh/datamodel/datasets.py (hunk 8 above)
Comment (jeromekelleher) [inline ga4gh/datamodel/datasets.py]: I think FileSystemDataset should be removed, and the SimulatedDataset doesn't do anything that a regular Dataset does, other than create a bunch of objects that it contains. Therefore, I think we should get rid of this polymorphism and just use plan Dataset everywhere, and decide what it contains at run-time.

Having said that, I do regret changing the name right now. It has led to a lot of diff noise!

File: ga4gh/datarepo.py (hunk 6 above)
Comment (jeromekelleher) [inline ga4gh/datarepo.py]: Good spot, done.

**Comment by jeromekelleher** [comment]
Thanks @dcolligan, good to hear you like it. I agree about disabling the tests, but you have to break some eggs some times. These will all be fixed before we merge to master. 

Updates over at #1079.
